File: ports/csharp/src/core/kernel/BcsObject.cs
using System;
using System.Collections.Generic;

namespace Bcs.Core.Kernel
{
    public class BcsObject : IDisposable
    {
        private BcsObject _parent;
        private readonly List<BcsObject> _children = new List<BcsObject>();
        private readonly object _lock = new object();
        private bool _isDisposed = false;

        public string ObjectName { get; set; } = string.Empty;

        public BcsObject(BcsObject parent = null)
        {
            if (parent != null)
            {
                SetParent(parent);
            }
        }

        public BcsObject Parent
        {
            get { lock (_lock) { return _parent; } }
        }

        public void SetParent(BcsObject newParent)
        {
            lock (_lock)
            {
                if (_parent != null && _parent != newParent)
                {
                    _parent.RemoveChild(this);
                }

                _parent = newParent;

                if (_parent != null)
                {
                    _parent.AddChild(this);
                }
            }
        }

        private void AddChild(BcsObject child)
        {
            lock (_lock)
            {
                if (!_children.Contains(child))
                {
                    _children.Add(child);
                }
            }
        }

        private void RemoveChild(BcsObject child)
        {
            lock (_lock)
            {
                _children.Remove(child);
            }
        }

        public virtual bool Event(BcsEvent e)
        {
            if (e.Type == EventType.Timer)
            {
                TimerEvent((BcsTimerEvent)e);
                return true;
            }
            return false;
        }

        protected virtual void TimerEvent(BcsTimerEvent e) { }
        protected virtual void ChildEvent(BcsChildEvent e) { }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (_isDisposed) return;

            if (disposing)
            {
                List<BcsObject> childrenCopy;
                lock (_lock)
                {
                    childrenCopy = new List<BcsObject>(_children);
                    _children.Clear();
                }

                foreach (var child in childrenCopy)
                {
                    child.Dispose();
                }

                SetParent(null);
            }

            _isDisposed = true;
        }

        ~BcsObject()
        {
            Dispose(false);
        }
    }
}

File: ports/java/src/core/kernel/BcsObject.java
package core.kernel;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.locks.ReentrantLock;

public class BcsObject implements AutoCloseable {
    private BcsObject parent;
    private final List<BcsObject> children = new CopyOnWriteArrayList<>();
    private final ReentrantLock lock = new ReentrantLock();
    private String objectName = "";
    private boolean isDestroyed = false;

    public BcsObject() {
        this(null);
    }

    public BcsObject(BcsObject parent) {
        if (parent != null) {
            setParent(parent);
        }
    }

    public String getObjectName() {
        lock.lock();
        try {
            return objectName;
        } finally {
            lock.unlock();
        }
    }

    public void setObjectName(String objectName) {
        lock.lock();
        try {
            this.objectName = objectName;
        } finally {
            lock.unlock();
        }
    }

    public BcsObject getParent() {
        lock.lock();
        try {
            return parent;
        } finally {
            lock.unlock();
        }
    }

    public void setParent(BcsObject newParent) {
        lock.lock();
        try {
            if (parent != null && parent != newParent) {
                parent.removeChild(this);
            }
            parent = newParent;
            if (parent != null) {
                parent.addChild(this);
            }
        } finally {
            lock.unlock();
        }
    }

    private void addChild(BcsObject child) {
        children.addIfAbsent(child);
    }

    private void removeChild(BcsObject child) {
        children.remove(child);
    }

    public boolean event(BcsEvent e) {
        if (e.getType() == BcsEvent.EventType.Timer) {
            timerEvent((BcsTimerEvent) e);
            return true;
        }
        return false;
    }

    protected void timerEvent(BcsTimerEvent e) { }
    protected void childEvent(BcsChildEvent e) { }

    @Override
    public void close() {
        lock.lock();
        try {
            if (isDestroyed) return;
            isDestroyed = true;

            for (BcsObject child : children) {
                child.close();
            }
            children.clear();
            setParent(null);
        } finally {
            lock.unlock();
        }
    }
}
